core/src/main/java/org/mule/runtime/core/internal/routing/UntilSuccessful.java

```diff
@@ -86,12 +113,23 @@ public CoreEvent process(CoreEvent event) throws MuleException {
     return processToApply(event, this);
   }
 
+  private RetryPolicyTemplate createRetryPolicyTemplate(CoreEvent event) {
+    Integer maxRetries =
+        (Integer) expressionManager.evaluate(this.maxRetries, DataType.NUMBER, NULL_BINDING_CONTEXT, event).getValue();
+    Integer millisBetweenRetries =
+        (Integer) expressionManager.evaluate(this.millisBetweenRetries, DataType.NUMBER, NULL_BINDING_CONTEXT, event).getValue();
+    return this.policyTemplatesCache.get(new Pair<>(millisBetweenRetries, maxRetries));
+  }
+
   @Override
   public Publisher<CoreEvent> apply(Publisher<CoreEvent> publisher) {
     return from(publisher)
         .flatMap(event -> Mono.from(processWithChildContextDontComplete(event, nestedChain, ofNullable(getLocation())))
-            .transform(p -> policyTemplate.applyPolicy(p, getRetryPredicate(), e -> {
-            }, getThrowableFunction(event), timer)));
+            .transform(p -> policyTemplate.orElseGet(() -> createRetryPolicyTemplate(event)).applyPolicy(p, getRetryPredicate(),
```

Comment:
extract the logic to fetch/build the retry policy to a provate method, for code readability